onnxruntime/python/tools/quantization/shape_inference.py

```diff
@@ -0,0 +1,261 @@
+        save_as_external_data: Saving an ONNX model to external data
+        all_tensors_to_one_file: Saving all the external data to one file
+        external_data_location: The file location to save the external file
+        external_data_size_threshold: The size threshold for external data
+    """
+    with tempfile.TemporaryDirectory(prefix="pre.quant.") as quant_tmp_dir:
+        temp_path = Path(quant_tmp_dir)
+        model = None
+
+        if not skip_symbolic_shape:
+            logger.info("Performing symbolic shape inference...")
+            model = SymbolicShapeInference.infer_shapes(
+                onnx.load(input_model_path),
+                int_max,
+                auto_merge,
+                guess_output_rank,
+                verbose,
+            )
+
+        if not skip_optimization:
+            # Use ORT optimizers (native code) to optimize model
+            if not skip_symbolic_shape:
+                # Need to save the inferenced model to file so as to run the optimizer
+                input_model_path = str(temp_path / "symbolic_shape_inferred.onnx")
+                onnx.save(model, input_model_path)
+                model = None
+
+            opt_model_path = str(temp_path / "optimized.onnx")
+            sess_option = onnxruntime.SessionOptions()
+            sess_option.optimized_model_filepath = opt_model_path
+            sess_option.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_BASIC
+            _ = onnxruntime.InferenceSession(input_model_path, sess_option, providers=["CPUExecutionProvider"])
+
+            input_model_path = opt_model_path
+
+        if not skip_onnx_shape:
+            # ONNX shape inference.
+            # According to docs, infer_shapes_path should be used for 2G+ models.
+            # If the skip optimization is specified, we could be dealing with a
+            # large model. So be on the safe side, save the model
+            if model is not None:
+                input_model_path = str(temp_path / "symbolic_shape_inferred.onnx")
+                if save_as_external_data:
+                    onnx.save_model(
+                        model,
+                        input_model_path,
+                        save_as_external_data=True,
+                        all_tensors_to_one_file=all_tensors_to_one_file,
+                        size_threshold=external_data_size_threshold,
+                        convert_attribute=False,
+                    )
+                else:
+                    onnx.save(model, input_model_path)
+                model = None
+
+            inferred_model_path = str(temp_path / "onnx_shape_inferred.onnx")
+            onnx.shape_inference.infer_shapes_path(input_model_path, inferred_model_path)
+            model = onnx.load(inferred_model_path)
+
+            # Add inference meta data to onnx model
+            metadata_props = {"onnx.infer": "onnxruntime.quant"}
```

Comment:
i think we can add another metadata {"onnx.quant.pre_process": True} and test it in quantization. If it is not set, we print out a warning message to ask user to call  quant_pre_process firstly.

And for optimize_model option in quantize_static, we can refine the comment like: this will be deprecated soon.